### PR TITLE
Support automatic host port assignment 

### PR DESCRIPTION
**Pull request by Zheaoli**
Support automatic host port assignment when people don't spec host port when they run a container 

Signed-off-by: Zheao.Li <me@manjusaka.me>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
What’s this?

**Comment by Zheaoli** [comment]
@AkihiroSuda I found some comment in codebase `automatic host port assignment is not supported yet (FIXME)`, So I make some code to support allocate host port when people try to run container

File: cmd/nerdctl/run.go [hunk 1]
```diff
@@ -128,6 +128,7 @@ func setCreateFlags(cmd *cobra.Command) {
 	cmd.Flags().StringSlice("dns", nil, "Set custom DNS servers")
 	// publish is defined as StringSlice, not StringArray, to allow specifying "--publish=80:80,443:443" (compatible with Podman)
 	cmd.Flags().StringSliceP("publish", "p", nil, "Publish a container's port(s) to the host")
+	cmd.Flags().Bool("auto-hostport", false, "Automatically allocate a host port for each port published")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run.go:131]: Docker doesn’t have such flag

File: pkg/portutil/portutil.go [hunk 2]
```diff
@@ -73,21 +75,30 @@ func ParseFlagP(s string) ([]gocni.PortMapping, error) {
...
 	}
-
+	if !needAutoAssignHostPort {
+		startHostPort, endHostPort, err = nat.ParsePortRange(hostPort)
+		if err != nil {
+			return nil, fmt.Errorf("invalid hostPort: %s", hostPort)
+		}
+	} else {
+		startHostPort, endHostPort = startPort, endPort
+		logrus.Warnf("auto-hostport is enabled, host port range is %d-%d", startHostPort, endHostPort)
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:100]: Docker assigns a random port

File: pkg/portutil/portutil.go [hunk 3]
```diff
@@ -45,7 +47,7 @@ func splitParts(rawport string) (string, string, string) {
 
 // ParseFlagP parse port mapping pair, like "127.0.0.1:3000:8080/tcp",
 // "127.0.0.1:3000-3001:8080-8081/tcp" and "3000:8080" ...
-func ParseFlagP(s string) ([]gocni.PortMapping, error) {
+func ParseFlagP(cmd *cobra.Command, s string) ([]gocni.PortMapping, error) {
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:50]: pkg/* has to be decoupled from cobra

File: pkg/portutil/portutil.go (hunk 2 above)
Comment (AkihiroSuda) [reply pkg/portutil/portutil.go:100]: And we need integration tests

File: pkg/portutil/portutil.go (hunk 3 above)
Comment (Zheaoli) [reply pkg/portutil/portutil.go:50]: Thanks for your review! I'll update my pr asap

File: pkg/portutil/portutil.go (hunk 2 above)
Comment (Zheaoli) [reply pkg/portutil/portutil.go:100]: I think if we need the same logic with docker. The nerdctl need an extra config to record which port has been allocated. 

Is it acceptable?

Comment (AkihiroSuda) [reply pkg/portutil/portutil.go:100]: >  The nerdctl need an extra config to record which port has been allocated.

No, https://github.com/lima-vm/lima/blob/943c90b13e38be32777b8f25be17c2491bb1421f/pkg/hostagent/hostagent.go#L181-L201

Comment (AkihiroSuda) [reply pkg/portutil/portutil.go:100]: You can also scan /proc/net/{tcp,tcp6} to find a free port

Comment (Zheaoli) [reply pkg/portutil/portutil.go:100]: Actually, in this first version of my thought before I make this PR, I have already thought about get a free port number by using listen "localhost:0". But I think it is not a good solution, here's my reason.

1. In the port range circumstance, we can't get a continuous port if we use the 'localhost:0' method.
2. In the Docker, the action of allocating port is trying to allocate the first free port in a specific port range which is not high frequency used in most circumstances (In Docker, the allocating range is from 

PTAL https://github.com/moby/moby/blob/master/libnetwork/portallocator/portallocator.go#L135-L176 and https://github.com/moby/moby/blob/master/libnetwork/portallocator/portallocator_unix.go#L7-L12

Comment (AkihiroSuda) [reply pkg/portutil/portutil.go:100]: Better to scan /proc/net/* then?

File: pkg/portutil/portutil.go [hunk 4]
```diff
@@ -73,21 +75,30 @@ func ParseFlagP(s string) ([]gocni.PortMapping, error) {
...
 	if hostPort == "" {
-		return nil, fmt.Errorf("automatic host port assignment is not supported yet (FIXME)")
-	}
-
-	startHostPort, endHostPort, err := nat.ParsePortRange(hostPort)
-	if err != nil {
-		return nil, fmt.Errorf("invalid hostPort: %s", hostPort)
+		commandConfig, err := cmd.Flags().GetBool("auto-hostport")
+		if err != nil || !commandConfig {
+			return nil, fmt.Errorf("no host port specified: %s", splitBySlash[0])
```
Comment (fahedouch) [inline pkg/portutil/portutil.go:82]: I assume that  `err` msg may be différent then `no host port is specified`. You should return err if not nil before checking the flag value

File: pkg/portutil/portutil.go (hunk 2 above)
Comment (Zheaoli) [reply pkg/portutil/portutil.go:100]: I'm not sure we should think about the compatibility of the Windows platform. Or we just need support auto port allocated for the Linux platform.

Comment (Zheaoli) [reply pkg/portutil/portutil.go:100]: ping @AkihiroSuda

Comment (AkihiroSuda) [reply pkg/portutil/portutil.go:100]: No need to cover windows now

Comment (Zheaoli) [reply pkg/portutil/portutil.go:100]: Got it.

**Comment by Zheaoli** [comment]
@AkihiroSuda I have already updated my PR. PTAL

File: cmd/nerdctl/run_network_linux_test.go [hunk 5]
```diff
@@ -140,6 +140,50 @@ func valuesOfMapStringString(m map[string]string) map[string]struct{} {
...
+			testContainerName := fmt.Sprintf("%s-%d", tID, i)
+			base := testutil.NewBase(t)
+			defer base.Cmd("rm", "-f", testContainerName).Run()
+			pFlag := tc.containerPort
+			cmd := base.Cmd("run", "-d",
+				"--name", testContainerName,
+				"-p", pFlag,
+				testutil.NginxAlpineImage)
+			if tc.runShouldSuccess {
+				cmd.AssertOK()
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_network_linux_test.go:177]: This is not enough. We have to assure that the port is allocated and connectable.

File: pkg/portutil/port_allocate.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
+//go:build !linux
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate.go:1]: Rename `port_allocate.go` to `port_allocate_linux.go`, and you can drop this line

File: pkg/portutil/port_allocate_linux.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
+//go:build linux
+// +build linux
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:2]: You can drop this line

File: pkg/portutil/port_allocate.go (hunk 6 above)
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate.go:1]: `port_allocate.go` -> `port_allocate_others.go`

File: pkg/portutil/port_allocate_linux.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+func parseHex(h string) (uint64, error) {
+	port, err := strconv.ParseInt(h, 16, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(port), nil
+}
+
+func getUsedPort(proto string) (map[uint64]uint8, error) {
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:80]: Looks like we are ignoring the IP and overestimating the set of the used port?

That is probably fine but that has to be documented in the code comments

Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:80]: This should take the content of /proc/net/* as `[]byte` argument, and should have unit tests

Comment (AkihiroSuda) [reply pkg/portutil/port_allocate_linux.go:80]: FYI: you may take some code from https://github.com/lima-vm/lima/tree/v0.8.3/pkg/guestagent/procnettcp (Apache License 2.0)

File: cmd/nerdctl/run_network_linux_test.go (hunk 5 above)
Comment (Zheaoli) [reply cmd/nerdctl/run_network_linux_test.go:177]: OK

File: pkg/portutil/port_allocate_linux.go (hunk 8 above)
Comment (Zheaoli) [reply pkg/portutil/port_allocate_linux.go:80]: Emm, I will add extra code for the IP case

Comment (Zheaoli) [reply pkg/portutil/port_allocate_linux.go:80]: Looks good to me! Thx for your advice

File: pkg/portutil/port_allocate_linux.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package portutil
+
+import (
+	"fmt"
+	"io/ioutil"
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:24]: ioutil is deprecated

**Comment by Zheaoli** [comment]
@AkihiroSuda Hello, Thanks for your patience. I have updated my PR. Would you mind to take a look again

**Comment by Zheaoli** [comment]
@AkihiroSuda Hello， would you mind taking a look at again?

**Comment by AkihiroSuda** [comment]
Not sure relevant, but CI can't be green after attempting 13 times

File: cmd/nerdctl/run_network_linux_test.go [hunk 10]
```diff
@@ -140,6 +143,78 @@ func valuesOfMapStringString(m map[string]string) map[string]struct{} {
...
+				result = cmd.Run()
+				stdoutContent = result.Stdout() + result.Stderr()
+				assert.Assert(cmd.Base.T, result.ExitCode == 0, stdoutContent)
+			} else {
+				cmd.AssertFail()
+				return
+			}
+			logrus.Infof("The output is %s", stdoutContent)
+			regexExpression := regexp.MustCompile(".*?the auto allocate port is from (?P<portNumber>\\d*?):80.*?")
+			match := regexExpression.FindStringSubmatch(stdoutContent)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_network_linux_test.go:196]: Why not use `nerdctl port` to check the port number?

File: pkg/portutil/portutil.go [hunk 11]
```diff
@@ -73,21 +74,25 @@ func ParseFlagP(s string) ([]gocni.PortMapping, error) {
...
 	if err != nil {
 		return nil, fmt.Errorf("invalid containerPort: %s", containerPort)
 	}
-
+	if hostPort == "" {
+		startHostPort, endHostPort, err = portAllocate(proto, ip, endPort-startPort+1)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Warnf("There is no hostPort has been spec in command, the auto allocate port is from %d:%d to %d:%d", startHostPort, startPort, endHostPort, endPort)
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:89]: logrus.Debugf

File: pkg/portutil/procnet/procnet.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	"strconv"
+	"strings"
+)
+
+type NetworkDetail struct {
+	LocalIP   net.IP
+	LocalPort uint64
+}
+
+func Parse(data []string) (results []NetworkDetail) {
```
Comment (AkihiroSuda) [inline pkg/portutil/procnet/procnet.go:32]: Please clarify the origin of this code and its license in the comment lines
https://github.com/lima-vm/lima/blob/v0.8.3/pkg/guestagent/procnettcp/procnettcp.go#L95-L137

File: pkg/portutil/procnet/procnet.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+func removeEmpty(array []string) (results []string) {
+	for _, i := range array {
+		if i != "" {
+			results = append(results, i)
+		}
+	}
+	return results
+}
+
+func ParseAddress(s string) (net.IP, uint16, error) {
```
Comment (AkihiroSuda) [inline pkg/portutil/procnet/procnet.go:57]: Please clarify the origin of this code and its license in the comment lines
https://github.com/lima-vm/lima/blob/v0.8.3/pkg/guestagent/procnettcp/procnettcp.go#L95-L137

File: pkg/portutil/procnet/procnetd_test.go [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+		"0: 00000000000000000000000000000000:0016 00000000000000000000000000000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 33825 1 0000000000000000 100 0 0 10 0",
+		"1: 00000000000000000000000000000000:006F 00000000000000000000000000000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 26772 1 0000000000000000 100 0 0 10 0",
+		"2: 00000000000000000000000000000000:0050 00000000000000000000000000000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 1210901 1 0000000000000000 100 0 0 10 0"}
+
+	entries := Parse(procNetTCP)
+	t.Log(entries)
+
+	assert.Check(t, net.IPv6zero.Equal(entries[0].LocalIP))
+	assert.Equal(t, uint64(22), entries[0].LocalPort)
+}
```
Comment (AkihiroSuda) [inline pkg/portutil/procnet/procnetd_test.go:66]: Please clarify the origin of this code and its license

File: cmd/nerdctl/run_network_linux_test.go (hunk 10 above)
Comment (Zheaoli) [reply cmd/nerdctl/run_network_linux_test.go:196]: Because we also need regex to get the endpoint from the `nerdctl port` output

**Comment by AkihiroSuda** [comment]
Please squash the commits

File: cmd/nerdctl/run_network_linux_test.go [hunk 15]
```diff
@@ -140,6 +142,80 @@ func valuesOfMapStringString(m map[string]string) map[string]struct{} {
 	return res
 }
 
+func TestRunPortWithNoHostPort(t *testing.T) {
+	testutil.DockerIncompatible(t)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/run_network_linux_test.go:146]: Why DockerIncompatible?

File: pkg/portutil/port_allocate_linux.go [hunk 16]
```diff
@@ -0,0 +1,79 @@
+//go:build linux
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:1]: You don't need `//go:build linux` for `*_linux.go`

File: pkg/portutil/port_allocate_linux.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+package portutil
+
+import (
+	"fmt"
+
+	"github.com/containerd/nerdctl/pkg/portutil/procnet"
+)
+
+const (
+	// This port range is compatible with Docker, FYI https://github.com/moby/moby/blob/master/libnetwork/portallocator/portallocator_unix.go#L7-L12
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:28]: > master...#L7-L12

Please use permalink with a commit hash or a tag

File: pkg/portutil/port_allocate_linux.go [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+		if filterFunc(s) {
+			ret = append(ret, s)
+		}
+	}
+	return
+}
+
+func portAllocate(protocol string, ip string, count uint64) (uint64, uint64, error) {
+	netprocData, err := procnet.ReadStatsFileData(protocol)
+	logrus.Warn("The file read complete")
```
Comment (AkihiroSuda) [inline pkg/portutil/port_allocate_linux.go:43]: Please remove this when this PR is in mergeable state

Comment (AkihiroSuda) [reply pkg/portutil/port_allocate_linux.go:43]: Marked the PR as a draft

Comment (Zheaoli) [reply pkg/portutil/port_allocate_linux.go:43]: I know it, I just test why we always failed in tonistiigi/binfmt now. I will remove it later

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**